File: webapp/src/app/features/shared/models/ride.ts
import { LocationDTO } from "./location";
import { VehicleType } from "./vehicle";
import {PassengerDTO} from './passenger';

export type RideStatus = 'PENDING' | 'ACCEPTED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED' | 'STOPPED';

export interface Ride {
    date: string;
    time: string;
    from: string;
    to: string;
    status: RideStatus;
    cancelledBy?: 'User' | 'Driver';
    panic: boolean;
    price: string;
}

export interface RideDTO {
  id: number;
  creatorEmail: string;
  driverEmail: string | null;
  status: RideStatus;
  startLocation: LocationDTO;
  endLocation: LocationDTO;
  waypoints: LocationDTO[];
  passengers: PassengerDTO[];
  passengerEmails: string[];
  vehicleType: VehicleType;
  basePrice: number;
  panicActivated: boolean;
  cancelledBy: string | null;
  scheduledTime: Date;
  isBabyFriendly: boolean;
  isPetsFriendly: boolean;
}

export interface RideDetailDTO {
  id: number;
  status: string;
  startAddress: string;
  endAddress: string;
  actualStartTime: string;
  actualEndTime: string;
  estimatedStartTime: string;
  estimatedEndTime: string;
  passengers: Array<{
    firstName: string;
    lastName: string;
    email: string;
  }>;
  totalPrice: number;
  cancelledBy: string | null;
  cancellationReason: string | null;
  cancellationTime: string | null;
  panicActivated: boolean;
  panicActivatedBy: string | null;
  panicActivatedAt: string | null;
  stoppedLocation: string | null;
  stoppedAt: string | null;
  inconsistencies: Array<{
    rideId: number;
    passengerId: number;
    passengerName: string;
    description: string;
    createdAt: string;
  }>;
}

File: backend/src/main/java/com/uberplus/backend/model/Ride.java
package com.uberplus.backend.model;

import com.uberplus.backend.model.enums.RideStatus;
import com.uberplus.backend.model.enums.VehicleType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "rides")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Ride {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creator_id", nullable = false)
    private Passenger creator;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "driver_id")
    private Driver driver;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RideStatus status = RideStatus.PENDING;

    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
    @JoinColumn(name = "start_location_id", nullable = false)
    private Location startLocation;

    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
    @JoinColumn(name = "end_location_id", nullable = false)
    private Location endLocation;

    @ManyToMany(cascade = CascadeType.PERSIST)
    @JoinTable(
            name = "ride_waypoints",
            joinColumns = @JoinColumn(name = "ride_id"),
            inverseJoinColumns = @JoinColumn(name = "location_id")
    )
    @OrderColumn(name = "order_index")
    private List<Location> waypoints = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "ride_passengers",
            joinColumns = @JoinColumn(name = "ride_id"),
            inverseJoinColumns = @JoinColumn(name = "passenger_id")
    )
    private List<Passenger> passengers = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private VehicleType vehicleType;

    @Column(nullable = false)
    private boolean babyFriendly;

    @Column(nullable = false)
    private boolean petsFriendly;

    @Column
    private LocalDateTime scheduledTime;

    @Column
    private LocalDateTime estimatedStartTime;

    @Column
    private LocalDateTime estimatedEndTime;

    @Column
    private LocalDateTime actualStartTime;

    @Column
    private LocalDateTime actualEndTime;

    @Column
    private Double distanceKm;

    @Column
    private Double basePrice;

    @Column
    private Double totalPrice;

    @Column
    private String cancelledBy;

    @Column
    private String cancellationReason;

    @Column
    private LocalDateTime cancellationTime;

    @Column(nullable = false)
    private boolean panicActivated = false;

    @Column
    private String panicActivatedBy;

    @Column
    private LocalDateTime panicActivatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stopped_location_id")
    private Location stoppedLocation;

    @Column
    private LocalDateTime stoppedAt;

    @Column(nullable = false)
    private LocalDateTime createdAt;

}

